sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuKeyBatchingIterator.scala

```diff
@@ -115,100 +117,119 @@ class GpuKeyBatchingIterator(
     }
   }
 
-  private def concatPending(last: Option[Table] = None): ColumnarBatch = {
-    var peak = 0L
-    try {
-      withResource(new NvtxWithMetrics("concat pending", NvtxColor.CYAN, concatTime)) { _ =>
-        withResource(mutable.ArrayBuffer[Table]()) { toConcat =>
-          while (pending.nonEmpty) {
-            withResource(pending.dequeue()) { spillable =>
+  private def concatPending(last: Option[SpillableColumnarBatch] = None): ColumnarBatch = {
+    val spillableBuffers = new ArrayBuffer[SpillableColumnarBatch]()
+    while (pending.nonEmpty) {
+      spillableBuffers.append(pending.dequeue())
+    }
+    pendingSize = 0
+    last.foreach { lastSpill =>
+      spillableBuffers.append(lastSpill)
+    }
+    RmmRapidsRetryIterator.withRetryNoSplit(spillableBuffers) { spillableBuffers =>
+      var peak = 0L
+      try {
+        withResource(new NvtxWithMetrics("concat pending", NvtxColor.CYAN, concatTime)) { _ =>
+          withResource(mutable.ArrayBuffer[Table]()) { toConcat =>
+            spillableBuffers.foreach { spillable =>
               withResource(spillable.getColumnarBatch()) { cb =>
                 peak += GpuColumnVector.getTotalDeviceMemoryUsed(cb)
                 toConcat.append(GpuColumnVector.from(cb))
               }
             }
-          }
-          pendingSize = 0
-          last.foreach { lastTab =>
-            peak += GpuColumnVector.getTotalDeviceMemoryUsed(lastTab)
-            toConcat.append(lastTab)
-          }
-          if (toConcat.length > 1) {
-            withResource(Table.concatenate(toConcat: _*)) { concated =>
-              peak += GpuColumnVector.getTotalDeviceMemoryUsed(concated)
-              GpuColumnVector.from(concated, types)
+
+            if (toConcat.length > 1) {
+              withResource(Table.concatenate(toConcat: _*)) { concated =>
+                peak += GpuColumnVector.getTotalDeviceMemoryUsed(concated)
+                GpuColumnVector.from(concated, types)
+              }
+            } else if (toConcat.nonEmpty) {
+              GpuColumnVector.from(toConcat.head, types)
+            } else {
+              // We got nothing but have to do something
+              GpuColumnVector.emptyBatchFromTypes(types)
             }
-          } else if (toConcat.nonEmpty) {
-            GpuColumnVector.from(toConcat.head, types)
-          } else {
-            // We got nothing but have to do something
-            GpuColumnVector.emptyBatchFromTypes(types)
           }
         }
+      } finally {
+        peakDevMemory.set(Math.max(peakDevMemory.value, peak))
+      }
+    }
+  }
+
+  private[this] def processAndCloseFinalBatch(cb: ColumnarBatch): ColumnarBatch = {
+    val scb = SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+    // concatPending will close scb
+    concatPending(Some(scb))
+  }
+
+  private[this] def splitAndCloseBatch(cb: ColumnarBatch): Option[ColumnarBatch] = {
```

Comment:
nit it would be nice to add a function header here that explains why the return could be None.